tests/e2e-kubernetes/testsuites/credentials.go

```diff
@@ -0,0 +1,256 @@
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+const (
+	iamRoleS3FullAccess     = "arn:aws:iam::aws:policy/AmazonS3FullAccess"
+	iamRoleS3ReadOnlyAccess = "arn:aws:iam::aws:policy/AmazonS3ReadOnlyAccess"
+)
+
+type s3CSICredentialsTestSuite struct {
+	tsInfo storageframework.TestSuiteInfo
+}
+
+func InitS3CSICredentialsTestSuite() storageframework.TestSuite {
+	return &s3CSICredentialsTestSuite{
+		tsInfo: storageframework.TestSuiteInfo{
+			Name: "credentials",
+			TestPatterns: []storageframework.TestPattern{
+				storageframework.DefaultFsPreprovisionedPV,
+			},
+		},
+	}
+}
+
+func (t *s3CSICredentialsTestSuite) GetTestSuiteInfo() storageframework.TestSuiteInfo {
+	return t.tsInfo
+}
+
+func (t *s3CSICredentialsTestSuite) SkipUnsupportedTests(_ storageframework.TestDriver, pattern storageframework.TestPattern) {
+	if pattern.VolType != storageframework.PreprovisionedPV {
+		e2eskipper.Skipf("Suite %q does not support %v", t.tsInfo.Name, pattern.VolType)
+	}
+}
+
+func (t *s3CSICredentialsTestSuite) DefineTests(driver storageframework.TestDriver, pattern storageframework.TestPattern) {
+	type local struct {
+		resources []*storageframework.VolumeResource
+		config    *storageframework.PerTestConfig
+	}
+	var l local
+
+	f := framework.NewFrameworkWithCustomTimeouts(NamespacePrefix+"credentials", storageframework.GetDriverTimeouts(driver))
+	f.NamespacePodSecurityLevel = admissionapi.LevelBaseline
+
+	cleanup := func(ctx context.Context) {
+		var errs []error
+		for _, resource := range l.resources {
+			errs = append(errs, resource.CleanupResource(ctx))
+		}
+		framework.ExpectNoError(errors.NewAggregate(errs), "while cleanup resource")
+	}
+	ginkgo.BeforeEach(func(ctx context.Context) {
+		l = local{}
+		l.config = driver.PrepareTest(ctx, f)
+		ginkgo.DeferCleanup(cleanup)
+	})
+	toWrite := 1024 // 1KB
+
+	expectWriteToSucceed := func(pod *v1.Pod) string {
+		seed := time.Now().UTC().UnixNano()
+		path := "/mnt/volume1/file.txt"
+		ginkgo.By(fmt.Sprintf("checking writing to %s", path))
+		checkWriteToPath(f, pod, path, toWrite, seed)
+		return path
+	}
+
+	expectReadToSucceed := func(pod *v1.Pod, path string) {
+		seed := time.Now().UTC().UnixNano()
+		ginkgo.By(fmt.Sprintf("checking writing to %s", path))
+		checkReadFromPath(f, pod, path, toWrite, seed)
+	}
+
+	expectWriteToFail := func(pod *v1.Pod) {
+		seed := time.Now().UTC().UnixNano()
+		path := "/mnt/volume1/file.txt"
+		ginkgo.By(fmt.Sprintf("checking if writing to %s fails", path))
+		checkWriteToPathFails(f, pod, path, toWrite, seed)
+	}
+
+	ginkgo.Context("Driver level", func() {
+		// TODO: Add tests for current functionality.
+	})
+
+	ginkgo.Context("Pod level", func() {
+		ginkgo.Context("should use correct credentials", func() {
+			ginkgo.It("full access role", func(ctx context.Context) {
+				sa, deleteSA := createServiceAccount(ctx, f, "s3-csi-e2e-sa", annotateServiceAccountWithRole(iamRoleS3FullAccess))
```

Comment:
If we're creating service accounts dynamically here, we can probably undo the changes to the cluster generation script